app/helpers.go

```diff
@@ -19,80 +19,81 @@ func NewABCIStore(app abci.Application) *ABCIStore {
 
 // Get will query for exactly one value over the abci store.
 // This can be wrapped with a bucket to reuse key/index/parse logic
-func (a *ABCIStore) Get(key []byte) []byte {
+func (a *ABCIStore) Get(key []byte) ([]byte, error) {
 	query := a.app.Query(abci.RequestQuery{
 		Path: "/",
 		Data: key,
 	})
 	// if only the interface supported returning errors....
 	if query.Code != 0 {
-		panic(query.Log)
+		return nil, errors.Wrap(errors.ErrDatabase, query.Log)
 	}
 	var value ResultSet
 	if err := value.Unmarshal(query.Value); err != nil {
-		panic(errors.Wrap(err, "unmarshal result set"))
+		return nil, errors.Wrapf(errors.ErrInvalidState, "unmarshal result set: %v", err.Error())
 	}
 	if len(value.Results) == 0 {
-		return nil
+		return nil, nil
 	}
 	// TODO: assert error if len > 1 ???
-	return value.Results[0]
+	return value.Results[0], nil
 }
 
 // Has returns true if the given key in in the abci app store
-func (a *ABCIStore) Has(key []byte) bool {
-	return len(a.Get(key)) > 0
+func (a *ABCIStore) Has(key []byte) (bool, error) {
+	got, err := a.Get(key)
+	if err != nil {
+		return false, err
+	}
+	return len(got) > 0, nil
 }
 
 // Iterator attempts to do a range iteration over the store,
 // We only support prefix queries in the abci server for now.
 // This client only supports listing everything...
-func (a *ABCIStore) Iterator(start, end []byte) weave.Iterator {
+func (a *ABCIStore) Iterator(start, end []byte) (weave.Iterator, error) {
 	// TODO: support all prefix searches (later even more ranges)
 	// look at orm/query.go:prefixRange for an idea how we turn prefix->iterator,
 	// we should detect this case and reverse it so we can serialize over abci query
 	if start != nil || end != nil {
-		panic("iterator only implemented for entire range")
+		return nil, errors.Wrap(errors.ErrDatabase, "iterator only implemented for entire range")
 	}
 
 	query := a.app.Query(abci.RequestQuery{
 		Path: "/?prefix",
 		Data: nil,
 	})
 	if query.Code != 0 {
-		panic(query.Log)
+		return nil, errors.Wrap(errors.ErrDatabase, query.Log)
 	}
 	models, err := toModels(query.Key, query.Value)
 	if err != nil {
-		panic(errors.Wrap(err, "cannot convert to model"))
+		return nil, errors.Wrapf(errors.ErrInvalidState, "cannot convert to model: %v", err.Error())
```

Comment:
`toModels` returns an `errors.Error` instance. Would it be better to wrap it instead of change to `ErrInvalidState`?

Reply:
Good point, I was getting very ambitious about not returning raw errors. will fix